replace ReactDOM.render() with createRoot() for React 18 compatibility

BurakParsAydin opened the pull request:
Fixes #7074 


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

BurakParsAydin committed:
replace ReactDOM.render() with createRoot() for React 18 compatibility

BurakParsAydin commented:
I have reviewed additional files and found that there are still more that need to be updated. Please hold off on merging the pull request until I have addressed these as well.

BurakParsAydin committed:
replace ReactDOM.render() with createRoot() for React 18 compatibility

BurakParsAydin committed:
Add changelog file

BurakParsAydin commented:
I have updated the other files as well. However, there are two additional files that still need to be updated. I have already attempted to update them, but an error occurs. The file paths are as follows:
`/client/order/order-status-change-strategies/index.tsx`
`/client/overview/task-list/tasks/go-live-task.tsx`

dmvrtx commented:
It appears that external imports should be updated too, to import `createRoot` as suggested [in the update path](https://react.dev/blog/2022/03/08/react-18-upgrade-guide)

BurakParsAydin commented:
@dmvrtx Just to clarify, would you like me to update React to version 18 in the package.json file as well? We are currently using version 17.0.2, and I want to ensure everything is consistent with the recommended changes. Thank you!

dmvrtx commented:
I suggest fixing the deprecations in this pull request and bumping React version in a follow-up.

BurakParsAydin commented:
@dmvrtx Thanks for your guidance. I've updated the React version in [here](https://github.com/Automattic/woocommerce-payments/pull/10475). I'll work on this PR as well.

BurakParsAydin committed:
fix: properly import createRoot from react-dom/client

BurakParsAydin commented:
@dmvrtx FYI, I have updated the files as per your suggestion.

Additionally, I have mentioned two more files above. Would you like me to update them as well?
Before proceeding, I believe I should first update the devDependencies [here](https://github.com/Automattic/woocommerce-payments/blob/develop/package.json#L108-L108)
> I have updated the other files as well. However, there are two additional files that still need to be updated. I have already attempted to update them, but an error occurs. The file paths are as follows: `/client/order/order-status-change-strategies/index.tsx` `/client/overview/task-list/tasks/go-live-task.tsx`

dmvrtx commented:
Thank you @BurakParsAydin 

> Before proceeding, I believe I should first update the devDependencies [here](https://github.com/Automattic/woocommerce-payments/blob/develop/package.json#L108-L108)

Sorry for not being clear in the comment on #10475. Yes, I agree, we need to update dependencies, both dev and production in this pull request. #10475 can then be closed.

Let's implement all the required changes in this one pull request.

BurakParsAydin commented:
@dmvrtx No worries at all, and I really appreciate your response!

I went ahead and updated the devDependencies to:
`"@types/react": "18.2.0", "@types/react-dom": "18.2.0",` but unfortunately, I encountered quite a few errors after making the update.I've attached a screenshot for reference. 

I've attached a screenshot for reference. Would it be alright to leave this as it is for now and address the remaining updates and fixes in a separate issue? Let me know your thoughts. I really appreciate your time!

![Screenshot from 2025-03-07 23-01-17](https://github.com/user-attachments/assets/2f9c7ea1-421a-4c7d-8315-311fed51dbf5)

dmvrtx commented:
> but unfortunately, I encountered quite a few errors after making the update.I've attached a screenshot for reference.

Thank you for looking into this, @BurakParsAydin. You might have got to find the reason why this migration hasn't been done yet. Those errors have to be fixed before the update can be merged into the production, and, seeing that the `createRoot` part of the change cannot be added without upgrading React, everything should be handled in this one PR.

If you see an opportunity to split it somehow so that each change will be passing tests clearly, it'll be welcome. But it is also okay to make everything in one PR, especially for version upgrades.

BurakParsAydin commented:
@dmvrtx Thank you for the update! I will try to proceed with this PR and apply all necessary updates here. Let's see how it goes.

BurakParsAydin committed:
chore(deps): upgrade React, related packages, and TypeScript

BurakParsAydin commented:
@dmvrtx Quick update, I've upgraded React and its related packages (`react`, `react-dom`, `react-test-renderer`, and `@types/react`) to their latest versions. During this process, I also had to update TypeScript, as the previous version was outdated and incompatible with the latest `@types/react`.

After performing the update, I ran `npx tsc --noEmit` to verify type correctness, and I encountered a number of TypeScript errors across multiple files. The errors are primarily related to missing or incompatible children props in various components and some implicit any type issues. Here are some key examples:

`TS2559: Type '{ children: Element; }' has no properties in common with type 'IntrinsicAttributes & PageProps'.`
`TS2339: Property 'children' does not exist on type 'SomeComponentProps'.`
`TS7006: Parameter 'someVariable' implicitly has an 'any' type.`
`TS2322: Type '{ children: ReactNode; className: string; }' is not assignable to type 'IntrinsicAttributes & SomeProps'.`

*Given these issues, I’d like to confirm the next steps. If you’d like, I can proceed with investigating and resolving these errors to ensure full compatibility with the updated dependencies. Let me know how you’d like to move forward*



---


Here are the complete files that have errors:
     1  client/capital/index.tsx:212
     1  client/card-readers/index.tsx:20
     1  client/components/amount-input/index.tsx:26
     1  client/components/cancel-authorization-button/index.tsx:25
     1  client/components/capture-authorization-button/index.tsx:25
     1  client/components/custom-select-control/index.tsx:133
     1  client/components/inline-label-select/index.tsx:157
     1  client/components/page/index.tsx:24
     1  client/components/payment-activity/index.tsx:160
     1  client/components/payment-activity/survey/context.tsx:66
     4  client/components/payment-activity/survey/test/index.test.tsx:33
     1  client/components/payment-methods-checkboxes/payment-method-checkbox.tsx:85
     1  client/components/pill/index.tsx:29
     1  client/components/tip-box/index.tsx:17
     1  client/components/tip-box/test/index.tsx:15
     1  client/components/woopay/index.tsx:49
     1  client/connect-account-page/index.tsx:471
     1  client/connect-account-page/info-notice-modal.tsx:30
     1  client/deposits/details/index.tsx:318
     1  client/deposits/index.tsx:151
     1  client/deposits/list/index.tsx:287
     1  client/disputes/index.tsx:449
     1  client/disputes/redirect-to-transaction-details/index.tsx:41
     1  client/documents/index.tsx:15
     1  client/documents/list/index.tsx:238
     1  client/onboarding/context.tsx:36
     1  client/onboarding/form.tsx:28
     5  client/onboarding/index.tsx:39
     2  client/onboarding/kyc/index.tsx:52
     1  client/onboarding/step.tsx:22
     1  client/onboarding/steps/test/business-details.tsx:176
     2  client/onboarding/test/context.tsx:48
     1  client/onboarding/test/form.tsx:89
     5  client/overview/task-list/tasks.tsx:120
     1  client/overview/task-list/tasks/add-apms-task.tsx:30
     5  client/overview/task-list/tasks/dispute-task.tsx:97
     1  client/overview/task-list/tasks/go-live-task.tsx:44
     1  client/overview/task-list/tasks/reconnect-task.tsx:30
     1  client/overview/task-list/tasks/update-business-details-task.tsx:148
     2  client/payment-details/payment-details/index.tsx:44
     2  client/payment-details/summary/index.tsx:460
     1  client/settings/express-checkout-settings/file-upload.tsx:103
     1  client/settings/express-checkout/woopay-item.tsx:186
     1  client/settings/fraud-protection/advanced-settings/allow-countries-notice.tsx:62
     2  client/settings/fraud-protection/advanced-settings/cards/address-mismatch.tsx:15
     2  client/settings/fraud-protection/advanced-settings/cards/avs-mismatch.tsx:16
     3  client/settings/fraud-protection/advanced-settings/cards/cvc-verification.tsx:21
     3  client/settings/fraud-protection/advanced-settings/cards/international-ip-address.tsx:24
     2  client/settings/fraud-protection/advanced-settings/cards/ip-address-mismatch.tsx:17
     5  client/settings/fraud-protection/advanced-settings/cards/order-items-threshold.tsx:122
     5  client/settings/fraud-protection/advanced-settings/cards/purchase-price-threshold.tsx:131
     3  client/settings/fraud-protection/advanced-settings/index.tsx:97
     1  client/settings/fraud-protection/advanced-settings/rule-card-notice.tsx:24
     3  client/settings/fraud-protection/advanced-settings/rule-card.tsx:22
     1  client/settings/fraud-protection/advanced-settings/rule-description.tsx:12
     1  client/settings/fraud-protection/advanced-settings/rule-toggle.tsx:65
     4  client/settings/fraud-protection/advanced-settings/test/rule-card-notice.test.tsx:13
     1  client/settings/fraud-protection/advanced-settings/test/rule-card.test.tsx:13
     1  client/settings/fraud-protection/advanced-settings/test/rule-description.test.tsx:13
     4  client/settings/fraud-protection/advanced-settings/test/rule-toggle.test.tsx:63
     1  client/settings/fraud-protection/index.tsx:20
     1  client/transactions/blocked/index.tsx:170
     1  client/transactions/index.tsx:109
     1  client/transactions/list/index.tsx:770
     1  client/transactions/risk-review/index.tsx:171
     1  client/transactions/uncaptured/index.tsx:253

dmvrtx commented:
Hi @BurakParsAydin. Once again, thanks for looking into how to make it possible. I think the best path is to fix those notices and elevate the quality of the codebase with this change. 

What is worth trying is to see if any of those reports can be fixed with the tools like [`ts-autofix`](https://github.com/ian-craig/ts-autofix) so that the main focus can be on the more complex changes.

BurakParsAydin commented:
Just a quick update - I attempted to use `ts-autofix` to speed up the TypeScript cleanups mentioned above, but it turns out the tool is not compatible with the current version of TypeScript used in the project.

It relies on internal APIs that no longer exist in newer TypeScript releases, which causes it to fail during execution. Rather than downgrading TypeScript (which could lead to other issues), I’ll go ahead and fix the issues manually, one by one.

BurakParsAydin committed:
resolve TypeScript errors to ensure compatibility with TypeScript v5.8.2

BurakParsAydin commented:
@dmvrtx FYI, I’ve completed the implementation, and the files are now ready for your review at your convenience. Please feel free to take a look, and let me know if you have any questions or require any clarifications.

frosso committed:
fix: tokenized ECE to honor woocommerce_tax_display_cart (#10662)

dmvrtx commented:
@BurakParsAydin , thank you for working on this. Let's first fix the pipelines, there are failing JS tests with the following error:

```
    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    /home/runner/work/woocommerce-payments/woocommerce-payments/node_modules/cheerio/dist/browser/index.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){export { contains, merge } from './static.js';
                                                                                      ^^^^^^

    SyntaxError: Unexpected token 'export'
```

I don't see that `cheerio` package being a direct dependency, so we might need to figure out first there it comes from as an indirect dependency and if it can be solved by upgrading that package. Or, if there are subtle changes between version upgrades which require additional settigns on the repository level.

alefesouza committed:
Update WooPay order icon (#10666)

BurakParsAydin commented:
@dmvrtx Thanks for your review. When I ran npm ls cheerio, the following output was returned:

> woocommerce-payments@9.1.0 /home/burak/Desktop/woocommerce-payments
> └─┬ @wordpress/jest-preset-default@8.1.2
>   └─┬ enzyme@3.11.0
>     └── cheerio@1.0.0

I did some research and found that enzyme@3.11.0 hasn’t had a stable release since 2019, and its dependency on cheerio@1.0.0 is locked. I also discovered that React 18 is not supported by Enzyme. Please see: https://stackoverflow.com/questions/72029189/does-enzyme-support-react-version-18

Would it be possible to remove Enzyme entirely?

dmvrtx commented:
> Would it be possible to remove Enzyme entirely?

@BurakParsAydin , we've looked into it and with #10674 it would go away and unblock your changes! Thank you for doing the heavy lifting with minor updates!

RadoslavGeorgiev committed:
Dev: Update to the latest jest-preset-default (#10674)

dmvrtx commented:
@BurakParsAydin, I tried to bring your branch up to date with the changes in the `develop` branch, but GitHub appears to be stuck with the "resolve conflicts" requirement. I suggest creating a new Pull Request from the forked repository with the same branch, as it shouldn't have any conflicts.

dmvrtx commented:
Okay, I think I managed to unstick the GitHub, and it now works. I've noticed a few errors in the tests locally. If those are reported by the checks, then those should be fixed, but at least we have packages at the right version.

BurakParsAydin commented:
@dmvrtx Thanks for reviewing and working on the branch. 
It seems that 7 checks were not successful. How should we proceed?

dmvrtx commented:
> It seems that 7 checks were not successful. How should we proceed?

If you look at the errors, they point to more places where `React.render()` should be replaced with `createRoot()`. Mostly in the tests as it seems. Those have to be updated. Can you check if it looks feasible? Otherwise, I'll take it to the team to discuss.

BurakParsAydin commented:
@dmvrtx Yes, I'm able to do it. Let me take a closer look and get started.

BurakParsAydin committed:
Remove unused bits

BurakParsAydin committed:
align React to v18.2.0 and update lockfile

BurakParsAydin committed:
migrate from ReactDOM.render to createRoot for React 18

BurakParsAydin commented:
@dmvrtx FYI, I’ve adjusted the files as requested, but we still have 6 failing tests. I took a quick look at the “JS linting and tests” job, and it seems the errors are related to some dependencies in node_modules. What do you suggest we do at this point?

dmvrtx commented:
@BurakParsAydin, I had a brief look at the JS linting and testing errors and most appear to fall in one of the two categories:

 - `React.render` is used in a test to verify that the resulting component contains or doesn't contain a specific element.
 - Snapshots have to be modified to include new hash values, probable linked to the updated component version.
 
The latter can be easily fixed with `npm run test:update-snapshots`. But the former is the same tedious manual work to replace all calls to `React.render` with `createRoot`. In the tests, it is often the case that `render` is imported from `React` and called without specifying the package.

BurakParsAydin committed:
update snapshots

BurakParsAydin commented:
@dmvrtx FYI, I ran `npm run test:update-snapshots` to update the snapshots before rebasing.
Then I rebased this branch onto the latest develop to resolve merge conflicts.
Some extra commits may appear in the history due to the rebase, but everything should now be aligned with develop.

Only left replacing `React.render` with `createRoot`

BurakParsAydin committed:
Update the packages

BurakParsAydin committed:
Update the snapshots

BurakParsAydin commented:
@dmvrtx I've addressed the React 18 compatibility issues within the WooPayments codebase, and from what I can see in the error logs, those parts are now working correctly. I also updated the relevant external packages in an effort to resolve the remaining issues.

However, _I’m currently stuck_, and am not quite sure how to move forward from here - the remaining test failures appear to be caused by incompatibilities in some of the external dependencies. It would be greatly appreciated if someone from the team could take a closer look at these remaining issues.

The affected files are listed below (there may be others, but these are the ones I’ve seen in the log):
```
client/disputes/filters/index.tsx
client/deposits/filters/test/index.js
client/settings/express-checkout-settings/__tests__/index.test.js
client/connect-account-page/test/index.test.tsx
client/documents/filters/test/index.test.tsx
client/components/phone-number-control/test/index.test.tsx
client/components/woopay/test/index.test.js
client/checkout/woopay/express-button/test/index.test.js
```

Thanks for the support! 🙏

dmvrtx commented:
@BurakParsAydin Thank you for all your work here. We're still looking into it to decide how to proceed further. I'll update on this pull request when there are more details.

dmvrtx committed:
Merge branch 'develop' into fix-7074

BurakParsAydin commented:
@dmvrtx Thanks for the update! I'm closely following this PR, so feel free to tag me when there's any news.

dmvrtx commented:
This update has been handled in a series of the separate pull requests. WooPayments now uses React 18.3. Thank you for the effort and patience with this one.
